File: Software/MenuPrototype/Incrementers/SpeedIncrementer.cs
using System;

namespace MenuPrototype
{
	partial class Program
	{
		public class SpeedIncrementer : Incrementer
		{
			int _speed = 0; // Come from mount
			int _ra = 0;
			int _dec = 0;
			int _back = 0;

			public override void onChange(string tag, int[] numbers, int index, int val)
			{
				base.onChange(tag, numbers, index, val);
				if (tag == "SPD")
				{
					_speed = numbers[0];
				}
				else if (tag == "RA")
				{
					_ra = numbers[0];
				}
				else if (tag == "DEC")
				{
					_dec = numbers[0];
				}
				else if (tag == "BACK")
				{
					_back = numbers[0];
				}
			}

			public override string getDisplay(string tag, int index, int val, string formatString = null)
			{
				if (tag == "SPD")
				{
					return string.Format("{0:0.0000}", 1.0 + 1.0 * val / 10000.0);
				}
				return val.ToString();
			}

			public override void getNumbers(string tag, int[] numbers)
			{
				if (tag == "SPD")
				{
					numbers[0] = _speed;
				}
				else if (tag == "RA")
				{
					numbers[0] = _ra;
				}
				else if (tag == "DEC")
				{
					numbers[0] = _dec;
				}
				else if (tag == "BACK")
				{
					numbers[0] = _back;
				}
			}

			public override void onSelect(EventArgs args)
			{
				Console.CursorTop = 4;
				if (args.source.getTag() == "SPD")
				{
					Console.WriteLine("Set speed factor to {0} ({1:0.0000})", _speed, 1.0 + 1.0 * _speed / 10000.0);
				}
				else if (args.source.getTag() == "RA")
				{
					Console.WriteLine("Set RA Steps to {0}", _ra);
				}
				else if (args.source.getTag() == "DEC")
				{
					Console.WriteLine("Set DEC Steps to {0}", _dec);
				}
				else if (args.source.getTag() == "BACK")
				{
					Console.WriteLine("Set Backlash asdjust to {0}", _back);
				}
			}
		}
	}
}

File: Software/Arduino code/OpenAstroTracker/speedIncrementer.cpp
#include "Configuration.hpp"
#include "lib/util/utils.hpp"
#include "lib/menu/controls/MainMenu.hpp"
#include "speedIncrementer.hpp"
#include "Mount.hpp"

SpeedAndStepIncrementer::SpeedAndStepIncrementer() : Incrementer()
{
}

void SpeedAndStepIncrementer::readValues()
{
    auto mount = Mount::instance();
    float speed = mount->getSpeedCalibration() - 1.0f;
    _speed = (int)(speed * 10000.0f);

    _raSteps = mount->getStepsPerDegree(RA_STEPS);
    _decSteps = mount->getStepsPerDegree(DEC_STEPS);
    _backlashSteps = mount->getBacklashCorrection();
}

void SpeedAndStepIncrementer::onChange(String tag, int *numbers, int index, int val)
{
    Incrementer::onChange(tag, numbers, index, val);
    if (tag.equals("SPD"))
    {
        _speed = numbers[0];
    }
    else if (tag.equals("RA"))
    {
        _raSteps = numbers[0];
    }
    else if (tag.equals("DEC"))
    {
        _decSteps = numbers[0];
    }
    else if (tag.equals("BACK"))
    {
        _backlashSteps = numbers[0];
    }
}

String SpeedAndStepIncrementer::getDisplay(String tag, int index, int val, String formatString)
{
    if (tag.equals("SPD"))
    {
        String(1.0 + 1.0 * val / 10000.0, 4);
    }

    return String(val);
}

void SpeedAndStepIncrementer::getNumbers(String tag, int *numbers)
{
    if (tag.equals("SPD"))
    {
        numbers[0] = _speed;
    }
    else if (tag.equals("RA"))
    {
        numbers[0] = _raSteps;
    }
    else if (tag.equals("DEC"))
    {
        numbers[0] = _decSteps;
    }
    else if (tag.equals("BACK"))
    {
        numbers[0] = _backlashSteps;
    }
}

void SpeedAndStepIncrementer::onSelect(EventArgs *args)
{
    auto mount = Mount::instance();
    if (args->getSource()->getTag().equals("SPD"))
    {
        mount->setSpeedCalibration(1.0 + _speed / 10000.0, true);
        args->getSource()->getMainMenu()->writeToLCD(0, 1, "Speed stored.");
        mount->delay(500);
    }
    else if (args->getSource()->getTag().equals("RA"))
    {
        mount->setStepsPerDegree(RA_STEPS, _raSteps);
        args->getSource()->getMainMenu()->writeToLCD(0, 1, "RA steps stored.");
        mount->delay(500);
    }
    else if (args->getSource()->getTag().equals("DEC"))
    {
        mount->setStepsPerDegree(DEC_STEPS, _decSteps);
        args->getSource()->getMainMenu()->writeToLCD(0, 1, "DEC steps stored.");
        mount->delay(500);
    }
    else if (args->getSource()->getTag().equals("BACK"))
    {
        mount->setBacklashCorrection(_backlashSteps);
        args->getSource()->getMainMenu()->writeToLCD(0, 1, "Backlash stored.");
        mount->delay(500);
    }
}
